samples: subsys: mgmt: updatehub: Add network manager and net interface overlays

nandojve opened the pull request:
This improves UpdateHub sample by using network manager. Now it is possible use the following technologies:

- Ethernet
- WIFI
- MODEM
- IEEE 802.15.4 (experimental)
- BLE IPSP (experimental)
- OpenThread (experimental)

The sample uses overlays to enable such network interface. It was reworked to add both WIFI connection block and network L4 events. The last one will trigger the option selected by user `manual` or `polling` mode.

At the UpdateHub core there is no relevant changes. Macros were reworked to improve the flash verification checks as suggested by @nvlsianpu on https://github.com/zephyrproject-rtos/zephyr/pull/27864#discussion_r483684298. The remaining changes are debug options to track update progress as debug logs.

Open as DNM because I still working on docs but I would like receive feedback on the sample code changes.
CC: @otavio

nandojve committed:
mgmt: updatehub: Kconfig: Drop NET_CONFIG_SETTINGS option

Remove NET_CONFIG_SETTINGS Kconfig option. The net settings should
be enabled from now at project config or at any overlay. This is
necessary to allow better control when application should start.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
samples: mgmt: updatehub: Add wifi support

Add WIFI overlay and connection request.  This allows users to evaluate
UpdateHub using any WIFI supported by Zephyr which supports UDP.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
mgmt: updatehub: Add debug logs

Current logs only prints default logs level. Add LOG_LEVEL at updatehub
to switch to another variation based on CONFIG_UPDATEHUB_LOG_LEVEL.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
samples: mgmt: updatehub: Add IEEE-802.15.4 support

Add IEEE 802.15.4 overlays.  This allows users to evaluate UpdateHub
using 6lowPAN.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
samples: mgmt: updatehub: Add BLE IPSP support

Add Bluetooth Low Energy IPSP overlays.  This allows users to evaluate
UpdateHub using BLE with IPSP supported by Zephyr.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
samples: mgmt: updatehub: Add network management

Add support to Network Management in UpdateHub sample.  This allows
start application based on network events, independent of the media
type selected by users.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
samples: mgmt: updatehub: Add modem support

Add MODEM and Arduino header overlays.  This allows users to evaluate
UpdateHub using any MODEM supported by Zephyr which supports UDP and
PPP connection.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
mgmt: updatehub: Clean SHA-256 macros

Clean SHA-256 macros that were left behind when added flash SHA-256
verification.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

nandojve committed:
samples: mgmt: updatehub: Add OpenThread support

Add OpenThread overlays.  This allows users to evaluate UpdateHub using
Thread Network.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

otavio commented on subsys/mgmt/updatehub/updatehub.c:
```diff
@@ -946,6 +953,8 @@ enum updatehub_response updatehub_update(void)
 
 	LOG_INF("Image flashed successfully, you can reboot now");
 
+	k_msleep(2000);
```
Why are you sleeping here?

nandojve replied:
When in polling mode need flush logs. Otherwise there is no "success" indicator after reboot. This keeps docs consistent with what users see.

otavio replied:
Couldn't UpdateHub use the immediate mode? As this is critical, it seems fine as fas we have a minimal number of log messages.

jukkar replied:
I do not recommend to set `CONFIG_LOG_IMMEDIATE` as it can slow down things if there is lot of logging output.
Just wondering why it is important that the user sees the info message? It could be that the info messages are not printed or there is no console where to print these. Shouldn't the status code be propagated up in the stack and eventually to application etc?

nandojve replied:
I agreed with Jukkar, CONFIG_LOG_IMMEDIATE is not the solution. The message isn't print because on polling mode after a successful upgrade system reboots immediately and buffered messages don't have time to be printed. Ideally a "LOG_FLUSH" is enough but I'm not sure if there is this kind of think available.

> Shouldn't the status code be propagated up in the stack and eventually to application etc?

In pooling mode UpdateHub takes control and reboot without communicate upper layers. When user wants control, it should execute **Manually** and check returned status. Polling mode is for simple apps like sensors that an eventual reset doesn't compromise system.

jukkar replied:
I wonder if `log_panic()` could be used here. Its documentation says that the logs are flushed. And as we reboot the device immediately, it should not matter if we enter the panic mode anyway.
```
/**
 * @brief Switch the logger subsystem to the panic mode.
 *
 * Returns immediately if the logger is already in the panic mode.
 *
 * @details On panic the logger subsystem informs all backends about panic mode.
 *          Backends must switch to blocking mode or halt. All pending logs
 *          are flushed after switching to panic mode. In panic mode, all log
 *          messages must be processed in the context of the call.
 */
__syscall void log_panic(void);
```

otavio replied:
Oh ... the panic kinda makes sense as we'll be forcing flushing and also informing other subsystems it'll reboot does sounds right.

nandojve replied:
I added LOG_PANIC() and tested with frdm_k64f.

nandojve committed:
samples: mgmt: updatehub: Update documentation

Update documentation using west enhancements and add sections related to
newer overlays.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

jukkar commented on samples/subsys/mgmt/updatehub/Kconfig:
```diff
@@ -20,4 +20,13 @@ config UPDATEHUB_POLLING
 endif #!UPDATEHUB_MANUAL
 
+if WIFI
+config UPDATEHUB_WIFI_SSID
```
In the networking side, if a sample app needs a kconfig option, I have added `SAMPLE` to the option name like `CONFIG_NET_SAMPLE_....`. This way there is no confusion what options are for the sample and what to the actual functionality.
So here I would recommend to name the option `config UPDATEHUB_SAMPLE_WIFI_SSID` etc.

jukkar commented on samples/subsys/mgmt/updatehub/src/main.c:
```diff
@@ -129,6 +129,12 @@ void main(void)
 		LOG_ERR("Error connecting to WiFi");
 		return;
 	}
+#elif defined(CONFIG_MODEM)
```
As this is GSM modem specific code, better to have `#elif defined(CONFIG_MODEM_GSM_PPP` here.

jukkar commented:
BTW, why do we have `samples/subsys/mgmt/updatehub/README.pdf` file in the sample dir, shouldn't the `README.rst` file be enough?

nandojve commented:
> BTW, why do we have `samples/subsys/mgmt/updatehub/README.pdf` file in the sample dir, shouldn't the `README.rst` file be enough?

It shouldn't be, already fixed!

nandojve commented:
I removed the DNM since PR have docs.

@jukkar , I think I addressed all your concerns . Can you revisit the PR, please?

nvlsianpu commented on samples/subsys/mgmt/updatehub/README.rst:
```diff
@@ -7,176 +7,341 @@ Overview
-	    --slot-size <image-slot-size> \
-	    --pad \
-        <path-to-zephyr.bin> signed.bin
+  west sign -t imgtool -d build/app -- --version 1.0.0 --pad
```
Build system might does that step automatically if  `CONFIG_MCUBOOT_SIGNATURE_KEY_FILE` is provided.
See also `CONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS` and `CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE`

nandojve replied:
My afraid is If we add this options right now it may confuse users. UpdateHub sample right now need instruct users about create two different image versions: 1) it will be used to flash the board, 2) is a different version to package for OTA. Not all parameters are equal for both images and for those boards that only have HEX enabled by default will be a problem when build/sign [2], because OTA needs the package with the BIN signed file not HEX.

My plan is enhance west to package UpdateHub. Than we can keep extra parameters internal for the second image sign and release to user the west flash/package without any confusion or mistake.

nvlsianpu replied:
ok, I just want be sure you are aware of.

nandojve commented on subsys/mgmt/updatehub/updatehub.c:
```diff
@@ -466,6 +466,11 @@ static void install_update_cb(void)
+	LOG_DBG("Flash: Address: 0x%08x, Size: %d, Flush: %d",
+		ctx.flash_ctx.stream.bytes_written,
+		response_packet.max_len - response_packet.offset,
+		ctx.downloaded_size == ctx.block.total_size);
```
True/False

nandojve commented on subsys/mgmt/updatehub/Kconfig:
```diff
@@ -15,7 +15,6 @@ menuconfig UPDATEHUB
 	select NET_SOCKETS
 	select NET_SOCKETS_POSIX_NAMES
 	select COAP
-	select NET_CONFIG_SETTINGS
```
We are moving to network manager L4 events and apps can start without network. UpdateHub core effectively will starts only after a L4 **connected** event.

jukkar commented on samples/subsys/mgmt/updatehub/README.rst:
```diff
@@ -210,38 +377,54 @@ And finally you can build the package by running:
 
 If everything is alright, it will print on the screen ``No update available``.
 
+For ``Pooling`` mode, the system will automatically register your device after
```
typo `s/Pooling/Polling/`

nandojve replied:
Fixed!

jukkar commented on samples/subsys/mgmt/updatehub/README.rst:
```diff
@@ -252,10 +435,12 @@ following command:
 updates, and then download the update package you've just created. If
 everything goes fine the message ``Image flashed successfully, you can reboot
-now`` will be printed on the terminal.
+now`` will be printed on the terminal.  If you are using the ``Pooling`` mode
```
same here, Pooling -> Polling, can you search and replace all of these (just in case there are more)

nandojve replied:
Fixed!

jukkar commented on samples/subsys/mgmt/updatehub/overlay-wifi.conf:
```diff
@@ -0,0 +1,23 @@
+# Disable default configs
+CONFIG_NET_L2_ETHERNET=n
+
+CONFIG_NET_DHCPV4=n
+
+CONFIG_NET_CONFIG_SETTINGS=n
```
BTW, if you disable DHCPv4 and static IP address settings in the overlay, the IP addresses need to be come from somewhere else. Who is setting the IP addresses to the network interface in this case?

nandojve replied:
Uhmm, I've been using ESP shield, default uses internal DHCP client. Which make sense since it is an offload driver.

See below the differences:

**CONFIG_NET_DHCPV4=n**
**CONFIG_NET_CONFIG_SETTINGS=n**
```
*** Booting Zephyr OS build zephyr-v2.4.0-3-gc2a0b0f50bdb  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.006,000] <inf> mcuboot: Primary image: magic=good, swap_type=0x2, copy_done=0x1, image_ok=0x1
[00:00:00.006,000] <inf> mcuboot: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.006,000] <inf> mcuboot: Boot source: none
[00:00:00.006,000] <inf> mcuboot: Swap type: none


uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-2239-gacd86316efd9  ***
[00:00:00.249,000] <inf> wifi_esp: Waiting for interface to come up
[00:00:00.622,000] <inf> wifi_esp: ESP Wi-Fi ready
[00:00:00.628,000] <inf> main: UpdateHub sample app started
[00:00:00.628,000] <inf> main: Confirming the boot image
[00:00:00.632,000] <inf> main: Network disconnected
[00:00:06.274,000] <inf> main: Network connected
[00:00:06.274,000] <inf> main: Starting UpdateHub polling mode
[00:00:06.274,000] <inf> updatehub: SHA-256 verification on download and from flash
uart:~$ net iface 

Interface 0x20000538 (IP Offload) [1]
=====================================
Link addr : XX:XX:XX:XX:XX:XX
MTU       : 2048
IPv4 unicast addresses (max 1):
        192.168.10.104 DHCP preferred infinite
IPv4 multicast addresses (max 1):
        <none>
IPv4 gateway : 192.168.10.1
IPv4 netmask : 255.255.255.0
uart:~$ 
```

**CONFIG_NET_DHCPV4=y**
**CONFIG_NET_CONFIG_SETTINGS=n**
```
*** Booting Zephyr OS build zephyr-v2.4.0-3-gc2a0b0f50bdb  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.006,000] <inf> mcuboot: Primary image: magic=good, swap_type=0x2, copy_done=0x1, image_ok=0x1
[00:00:00.006,000] <inf> mcuboot: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.006,000] <inf> mcuboot: Boot source: none
[00:00:00.006,000] <inf> mcuboot: Swap type: none


uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-2239-gacd86316efd9  ***
[00:00:00.250,000] <inf> wifi_esp: Waiting for interface to come up
[00:00:00.623,000] <inf> wifi_esp: ESP Wi-Fi ready
[00:00:00.629,000] <inf> main: UpdateHub sample app started
[00:00:00.629,000] <inf> main: Confirming the boot image
[00:00:00.633,000] <inf> main: Network disconnected
[00:00:06.274,000] <inf> main: Network connected
[00:00:06.274,000] <inf> main: Starting UpdateHub polling mode
[00:00:06.274,000] <inf> updatehub: SHA-256 verification on download and from flash
[00:00:08.642,000] <err> updatehub: Could not receive data
uart:~$ net iface 

Interface 0x20000540 (IP Offload) [1]
=====================================
Link addr : XX:XX:XX:XX:XX:XX
MTU       : 2048
IPv4 unicast addresses (max 1):
        192.168.10.104 DHCP preferred infinite
IPv4 multicast addresses (max 1):
        <none>
IPv4 gateway : 192.168.10.1
IPv4 netmask : 255.255.255.0
DHCPv4 lease time : 0
DHCPv4 renew time : 0
DHCPv4 server     : 0.0.0.0
DHCPv4 requested  : 0.0.0.0
DHCPv4 state      : disabled
DHCPv4 attempts   : 0
uart:~$
```

**CONFIG_NET_DHCPV4=n**
**CONFIG_NET_CONFIG_SETTINGS=y**
**CONFIG_NET_CONFIG_NEED_IPV4=y**
**CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.168.10.104"**
**CONFIG_NET_CONFIG_MY_IPV4_GW="192.168.10.1"**
```
*** Booting Zephyr OS build zephyr-v2.4.0-3-gc2a0b0f50bdb  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.006,000] <inf> mcuboot: Primary image: magic=good, swap_type=0x2, copy_done=0x1, image_ok=0x1
[00:00:00.006,000] <inf> mcuboot: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.006,000] <inf> mcuboot: Boot source: none
[00:00:00.006,000] <inf> mcuboot: Swap type: none


uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-2239-gacd86316efd9  ***
[00:00:00.249,000] <inf> wifi_esp: Waiting for interface to come up
[00:00:00.621,000] <inf> wifi_esp: ESP Wi-Fi ready
[00:00:00.628,000] <inf> main: UpdateHub sample app started
[00:00:00.628,000] <inf> main: Confirming the boot image
[00:00:00.632,000] <inf> main: Network connected
[00:00:00.632,000] <inf> main: Starting UpdateHub polling mode
[00:00:00.632,000] <inf> updatehub: SHA-256 verification on download and from flash
[00:00:07.592,000] <err> updatehub: Could not receive data
uart:~$ net iface 

Interface 0x20000550 (IP Offload) [1]
=====================================
Link addr : XX:XX:XX:XX:XX:XX
MTU       : 2048
IPv4 unicast addresses (max 1):
        192.168.10.104 manual preferred infinite
IPv4 multicast addresses (max 1):
        <none>
IPv4 gateway : 192.168.10.1
IPv4 netmask : 255.255.255.0
uart:~$ 
```

**CONFIG_NET_DHCPV4=y**
**CONFIG_NET_CONFIG_SETTINGS=y**
**CONFIG_NET_CONFIG_NEED_IPV4=y**
**CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.168.10.104"**
**CONFIG_NET_CONFIG_MY_IPV4_GW="192.168.10.1"**
```
*** Booting Zephyr OS build zephyr-v2.4.0-3-gc2a0b0f50bdb  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.006,000] <inf> mcuboot: Primary image: magic=good, swap_type=0x2, copy_done=0x1, image_ok=0x1
[00:00:00.006,000] <inf> mcuboot: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.006,000] <inf> mcuboot: Boot source: none
[00:00:00.006,000] <inf> mcuboot: Swap type: none


uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-2239-gacd86316efd9  ***
[00:00:00.246,000] <inf> wifi_esp: Waiting for interface to come up
[00:00:00.618,000] <inf> wifi_esp: ESP Wi-Fi ready
[00:00:00.624,000] <inf> main: UpdateHub sample app started
[00:00:00.624,000] <inf> main: Confirming the boot image
[00:00:00.629,000] <inf> main: Network connected
[00:00:00.629,000] <inf> main: Starting UpdateHub polling mode
[00:00:00.629,000] <inf> updatehub: SHA-256 verification on download and from flash
[00:00:07.584,000] <err> updatehub: Could not receive data
uart:~$ net iface 

Interface 0x20000558 (IP Offload) [1]
=====================================
Link addr : 5C:CF:7F:72:4D:A2
MTU       : 2048
IPv4 unicast addresses (max 1):
        192.168.10.104 overridable preferred infinite
IPv4 multicast addresses (max 1):
        <none>
IPv4 gateway : 192.168.10.1
IPv4 netmask : 255.255.255.0
DHCPv4 lease time : 0
DHCPv4 renew time : 0
DHCPv4 server     : 0.0.0.0
DHCPv4 requested  : 0.0.0.0
DHCPv4 state      : init
DHCPv4 attempts   : 0
uart:~$ 
```

jukkar replied:
Ok, makes sense.

jukkar reviewed (approved):
LGTM
